File: android/MyNotes/app/src/main/java/io/qzz/superstar1014/mynotes/HermesChatCrypto.java
package io.qzz.superstar1014.mynotes;

import org.json.JSONArray;
import org.json.JSONObject;

import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.UUID;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;

final class HermesChatCrypto {
  static final int MAX_ATTACHMENT_BYTES = 10 * 1024 * 1024;
  static final int MAX_AGENT_ATTACHMENT_BYTES = 512 * 1024 * 1024;
  static final String PRIVATE_ATTACHMENT_STORAGE = "r2-private-v1";
  private static final SecureRandom RANDOM = new SecureRandom();

  private final SecretKeySpec key;

  HermesChatCrypto(String encodedKey) {
    byte[] raw;
    try {
      raw = Base64.getUrlDecoder().decode(encodedKey == null ? "" : encodedKey.trim());
    } catch (IllegalArgumentException error) {
      throw new IllegalArgumentException("聊天密钥不是有效的 Base64URL。", error);
    }
    if (raw.length != 32) throw new IllegalArgumentException("聊天密钥必须正好包含 32 字节。");
    key = new SecretKeySpec(raw, "AES");
  }

  static String generateKey() {
    byte[] raw = new byte[32];
    RANDOM.nextBytes(raw);
    return encode(raw);
  }

  JSONObject encryptMessage(
    String sender,
    String text,
    JSONArray attachments,
    String messageId,
    long sentAt
  ) throws Exception {
    return encryptMessage(sender, text, attachments, messageId, sentAt, 0, false);
  }

  JSONObject encryptMessage(
    String sender,
    String text,
    JSONArray attachments,
    String messageId,
    long sentAt,
    long replaceSequence,
    boolean finalUpdate
  ) throws Exception {
    if (!"client".equals(sender) && !"agent".equals(sender)) {
      throw new IllegalArgumentException("消息发送者无效。");
    }
    JSONArray safeAttachments = attachments == null ? new JSONArray() : attachments;
    validateAttachmentList(safeAttachments, sender);
    JSONObject plaintext = new JSONObject()
      .put("text", text == null ? "" : text)
      .put("attachments", safeAttachments);
    if (replaceSequence > 0) {
      plaintext.put("replaceSeq", replaceSequence).put("final", finalUpdate);
    } else if (replaceSequence < 0) {
      throw new IllegalArgumentException("消息替换序号无效。");
    }
    return encryptPayload(sender, messageId, sentAt, plaintext);
  }

  JSONObject encryptInteractionResponse(
    String promptId,
    String optionId,
    String messageId,
    long sentAt
  ) throws Exception {
    requireUuid(promptId, "交互请求 id 无效。");
    if (!isInteractionToken(optionId, 64)) {
      throw new IllegalArgumentException("交互选项 id 无效。");
    }
    JSONObject plaintext = new JSONObject()
      .put("text", "")
      .put("attachments", new JSONArray())
      .put("interactionResponse", new JSONObject()
        .put("promptId", promptId)
        .put("optionId", optionId));
    return encryptPayload("client", messageId, sentAt, plaintext);
  }

  ChatMessage decryptMessage(JSONObject envelope, long sequence) throws Exception {
    if (envelope.optInt("v") != 1 || !"message".equals(envelope.optString("type"))) {
      throw new IllegalArgumentException("收到不支持的聊天消息。");
    }
    String sender = envelope.optString("sender");
    if (!"client".equals(sender) && !"agent".equals(sender)) {
      throw new IllegalArgumentException("聊天消息发送者无效。");
    }
    String messageId = envelope.getString("id");
    long sentAt = envelope.getLong("sentAt");
    JSONObject encrypted = envelope.getJSONObject("encrypted");
    if (!"A256GCM".equals(encrypted.optString("alg"))) {
      throw new IllegalArgumentException("不支持此聊天加密格式。");
    }
    byte[] plaintext = crypt(
      Cipher.DECRYPT_MODE,
      decode(encrypted.getString("ciphertext")),
      decode(encrypted.getString("nonce")),
      messageAad(messageId, sender, sentAt)
    );
    JSONObject payload = new JSONObject(new String(plaintext, StandardCharsets.UTF_8));
    JSONObject interaction = payload.optJSONObject("interaction");
    if (interaction != null) {
      if (!"agent".equals(sender)) {
        throw new IllegalArgumentException("仅 Hermes 可以发送交互操作。");
      }
      interaction = validateInteraction(interaction);
    }
    long replaceSequence = payload.has("replaceSeq") ? payload.getLong("replaceSeq") : 0;
    boolean finalUpdate = payload.has("final") && payload.getBoolean("final");
    if ((payload.has("replaceSeq") || payload.has("final"))
        && (replaceSequence < 1 || !payload.has("replaceSeq") || !payload.has("final"))) {
      throw new IllegalArgumentException("消息替换元数据无效。");
    }
    JSONArray attachments = payload.optJSONArray("attachments") == null
      ? new JSONArray()
      : payload.getJSONArray("attachments");
    validateAttachmentList(attachments, sender);
    return new ChatMessage(
      messageId,
      sender,
      sentAt,
      sequence,
      payload.optString("text"),
      attachments,
      interaction == null ? new JSONObject() : interaction,
      replaceSequence,
      finalUpdate
    );
  }

  private JSONObject encryptPayload(
    String sender,
    String messageId,
    long sentAt,
    JSONObject plaintext
  ) throws Exception {
    requireUuid(messageId, "消息 id 无效。");
    byte[] nonce = new byte[12];
    RANDOM.nextBytes(nonce);
    byte[] ciphertext = crypt(
      Cipher.ENCRYPT_MODE,
      plaintext.toString().getBytes(StandardCharsets.UTF_8),
      nonce,
      messageAad(messageId, sender, sentAt)
    );
    return new JSONObject()
      .put("v", 1)
      .put("type", "message")
      .put("id", messageId)
      .put("sender", sender)
      .put("sentAt", sentAt)
      .put("encrypted", new JSONObject()
        .put("alg", "A256GCM")
        .put("nonce", encode(nonce))
        .put("ciphertext", encode(ciphertext)));
  }

  EncryptedAttachment encryptAttachment(
    byte[] plaintext,
    String contentType,
    String filename
  ) throws Exception {
    if (plaintext == null || plaintext.length < 1 || plaintext.length > MAX_ATTACHMENT_BYTES) {
      throw new IllegalArgumentException("附件大小必须在 10 MiB 以内。");
    }
    String attachmentId = UUID.randomUUID().toString();
    String safeType = safeContentType(contentType);
    byte[] nonce = new byte[12];
    RANDOM.nextBytes(nonce);
    byte[] ciphertext = crypt(
      Cipher.ENCRYPT_MODE,
      plaintext,
      nonce,
      attachmentAad(attachmentId, safeType, plaintext.length)
    );
    JSONObject descriptor = new JSONObject()
      .put("id", attachmentId)
      .put("name", safeFilename(filename))
      .put("contentType", safeType)
      .put("plaintextBytes", plaintext.length)
      .put("nonce", encode(nonce));
    return new EncryptedAttachment(descriptor, ciphertext);
  }

  byte[] decryptAttachment(byte[] ciphertext, JSONObject descriptor) throws Exception {
    validateAttachmentDescriptor(descriptor, "client");
    String attachmentId = descriptor.getString("id");
    String contentType = safeContentType(descriptor.getString("contentType"));
    int plaintextBytes = descriptor.getInt("plaintextBytes");
    if (plaintextBytes < 1 || plaintextBytes > MAX_ATTACHMENT_BYTES) {
      throw new IllegalArgumentException("附件大小无效。");
    }
    byte[] plaintext = crypt(
      Cipher.DECRYPT_MODE,
      ciphertext,
      decode(descriptor.getString("nonce")),
      attachmentAad(attachmentId, contentType, plaintextBytes)
    );
    if (plaintext.length != plaintextBytes) {
      throw new IllegalArgumentException("附件完整性校验失败。");
    }
    return plaintext;
  }

  JSONObject encryptLocalSnapshot(JSONObject payload, String spaceId) throws Exception {
    byte[] nonce = new byte[12];
    RANDOM.nextBytes(nonce);
    byte[] ciphertext = crypt(
      Cipher.ENCRYPT_MODE,
      payload.toString().getBytes(StandardCharsets.UTF_8),
      nonce,
      localSnapshotAad(spaceId)
    );
    return new JSONObject()
      .put("v", 1)
      .put("alg", "A256GCM")
      .put("nonce", encode(nonce))
      .put("ciphertext", encode(ciphertext));
  }

  JSONObject decryptLocalSnapshot(JSONObject envelope, String spaceId) throws Exception {
    if (envelope.optInt("v") != 1 || !"A256GCM".equals(envelope.optString("alg"))) {
      throw new IllegalArgumentException("本地聊天缓存格式无效。");
    }
    byte[] plaintext = crypt(
      Cipher.DECRYPT_MODE,
      decode(envelope.getString("ciphertext")),
      decode(envelope.getString("nonce")),
      localSnapshotAad(spaceId)
    );
    return new JSONObject(new String(plaintext, StandardCharsets.UTF_8));
  }

  private byte[] crypt(int mode, byte[] input, byte[] nonce, byte[] aad) throws Exception {
    if (nonce.length != 12) throw new IllegalArgumentException("AES-GCM nonce 长度无效。");
    Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
    cipher.init(mode, key, new GCMParameterSpec(128, nonce));
    cipher.updateAAD(aad);
    return cipher.doFinal(input);
  }

  private static byte[] messageAad(String id, String sender, long sentAt) {
    return ("hermes-chat-message-v1|" + id + "|" + sender + "|" + sentAt)
      .getBytes(StandardCharsets.UTF_8);
  }

  private static byte[] attachmentAad(String id, String contentType, int bytes) {
    return ("hermes-chat-attachment-v1|" + id + "|" + contentType + "|" + bytes)
      .getBytes(StandardCharsets.UTF_8);
  }

  private static byte[] localSnapshotAad(String spaceId) {
    String normalized = spaceId == null
      ? ""
      : spaceId.trim().toLowerCase(java.util.Locale.ROOT);
    if (!normalized.matches("[a-z0-9][a-z0-9_-]{0,63}")) {
      throw new IllegalArgumentException("Hermes profile id 无效。");
    }
    return ("hermes-chat-local-cache-v1|" + normalized).getBytes(StandardCharsets.UTF_8);
  }

  private static String safeFilename(String value) {
    String filename = value == null ? "attachment" : value.replaceAll("[^\\p{L}\\p{N}._ -]", "_");
    filename = filename.replaceAll("^[ .]+|[ .]+$", "");
    if (filename.isEmpty()) filename = "attachment";
    return filename.length() > 120 ? filename.substring(0, 120) : filename;
  }

  private static String safeContentType(String value) {
    String contentType = value == null
      ? "application/octet-stream"
      : value.split(";", 2)[0].trim().toLowerCase(java.util.Locale.ROOT);
    if (contentType.length() > 100 || !contentType.matches("[a-z0-9!#$&^_.+\\-/]+")) {
      throw new IllegalArgumentException("附件类型无效。");
    }
    return contentType;
  }

  static boolean isPrivateAttachment(JSONObject descriptor) {
    return descriptor != null
      && PRIVATE_ATTACHMENT_STORAGE.equals(descriptor.optString("storage"));
  }

  static void validateAttachmentDescriptor(JSONObject descriptor, String sender) throws Exception {
    if (descriptor == null) throw new IllegalArgumentException("附件描述无效。");
    requireUuid(descriptor.getString("id"), "附件 id 无效。");
    String name = descriptor.getString("name");
    String contentType = descriptor.getString("contentType");
    if (name.isEmpty() || name.length() > 120 || !safeContentType(contentType).equals(contentType)) {
      throw new IllegalArgumentException("附件元数据无效。");
    }
    int plaintextBytes = descriptor.getInt("plaintextBytes");
    if (isPrivateAttachment(descriptor)) {
      String sha256 = descriptor.optString("sha256");
      if (!"agent".equals(sender)
          || plaintextBytes <= MAX_ATTACHMENT_BYTES
          || plaintextBytes > MAX_AGENT_ATTACHMENT_BYTES
          || !sha256.matches("[0-9a-f]{64}")
          || (!descriptor.optString("nonce").isEmpty())) {
        throw new IllegalArgumentException("Hermes 大附件描述无效。");
      }
      return;
    }
    String storage = descriptor.optString("storage");
    if ((!storage.isEmpty() && !"encrypted-v1".equals(storage))
        || plaintextBytes < 1
        || plaintextBytes > MAX_ATTACHMENT_BYTES
        || decode(descriptor.getString("nonce")).length != 12
        || !descriptor.optString("sha256").isEmpty()) {
      throw new IllegalArgumentException("加密附件描述无效。");
    }
  }

  private static void validateAttachmentList(JSONArray attachments, String sender) throws Exception {
    if (attachments.length() > 8) throw new IllegalArgumentException("一条消息最多包含 8 个附件。");
    for (int index = 0; index < attachments.length(); index++) {
      validateAttachmentDescriptor(attachments.getJSONObject(index), sender);
    }
  }

  private static JSONObject validateInteraction(JSONObject value) throws Exception {
    requireUuid(value.getString("id"), "交互请求 id 无效。");
    String kind = value.getString("kind");
    if (!kind.matches("approval|slash_confirm|clarify|model|choice")) {
      throw new IllegalArgumentException("交互类型无效。");
    }
    String state = value.getString("state");
    if (!state.matches("pending|resolved|expired|error")) {
      throw new IllegalArgumentException("交互状态无效。");
    }
    Object expires = value.get("expiresAt");
    if (!(expires instanceof Number) || ((Number) expires).longValue() < 0) {
      throw new IllegalArgumentException("交互过期时间无效。");
    }
    JSONArray options = value.getJSONArray("options");
    if (options.length() > 24 || ("pending".equals(state) && options.length() == 0)) {
      throw new IllegalArgumentException("交互选项数量无效。");
    }
    for (int index = 0; index < options.length(); index++) {
      JSONObject option = options.getJSONObject(index);
      if (!isInteractionToken(option.getString("id"), 64)) {
        throw new IllegalArgumentException("交互选项 id 无效。");
      }
      String label = option.getString("label");
      if (label.isEmpty() || label.length() > 120) {
        throw new IllegalArgumentException("交互选项文字无效。");
      }
      if (!option.optString("style", "default").matches("default|primary|danger|warning")) {
        throw new IllegalArgumentException("交互选项样式无效。");
      }
      for (String flag : new String[] {"selected", "disabled", "wide"}) {
        if (option.has(flag) && !(option.get(flag) instanceof Boolean)) {
          throw new IllegalArgumentException("交互选项状态无效。");
        }
      }
    }
    if (value.optString("stage").length() > 32
        || value.optString("status").length() > 500
        || value.optString("pageInfo").length() > 100) {
      throw new IllegalArgumentException("交互说明文字过长。");
    }
    return new JSONObject(value.toString());
  }

  private static void requireUuid(String value, String message) {
    try {
      if (!UUID.fromString(value).toString().equals(value.toLowerCase(java.util.Locale.ROOT))) {
        throw new IllegalArgumentException(message);
      }
    } catch (Exception error) {
      throw new IllegalArgumentException(message, error);
    }
  }

  private static boolean isInteractionToken(String value, int maximum) {
    return value != null
      && value.length() >= 1
      && value.length() <= maximum
      && value.matches("[A-Za-z0-9_-]+");
  }

  private static JSONObject copyObject(JSONObject value) {
    try {
      return new JSONObject(value == null ? "{}" : value.toString());
    } catch (Exception ignored) {
      return new JSONObject();
    }
  }

  private static String encode(byte[] value) {
    return Base64.getUrlEncoder().withoutPadding().encodeToString(value);
  }

  private static byte[] decode(String value) {
    return Base64.getUrlDecoder().decode(value);
  }

  static final class ChatMessage {
    final String id;
    final String sender;
    final long sentAt;
    long sequence;
    final String text;
    final JSONArray attachments;
    final JSONObject interaction;
    final long replaceSequence;
    final boolean finalUpdate;

    ChatMessage(
      String id,
      String sender,
      long sentAt,
      long sequence,
      String text,
      JSONArray attachments
    ) {
      this(id, sender, sentAt, sequence, text, attachments, new JSONObject(), 0, false);
    }

    ChatMessage(
      String id,
      String sender,
      long sentAt,
      long sequence,
      String text,
      JSONArray attachments,
      long replaceSequence,
      boolean finalUpdate
    ) {
      this(id, sender, sentAt, sequence, text, attachments, new JSONObject(), replaceSequence, finalUpdate);
    }

    ChatMessage(
      String id,
      String sender,
      long sentAt,
      long sequence,
      String text,
      JSONArray attachments,
      JSONObject interaction,
      long replaceSequence,
      boolean finalUpdate
    ) {
      this.id = id;
      this.sender = sender;
      this.sentAt = sentAt;
      this.sequence = sequence;
      this.text = text;
      this.attachments = attachments;
      this.interaction = copyObject(interaction);
      this.replaceSequence = replaceSequence;
      this.finalUpdate = finalUpdate;
    }
  }

  static final class EncryptedAttachment {
    final JSONObject descriptor;
    final byte[] ciphertext;

    EncryptedAttachment(JSONObject descriptor, byte[] ciphertext) {
      this.descriptor = descriptor;
      this.ciphertext = ciphertext;
    }
  }
}

File: macos/AIBuildNotes/Sources/NotesCore/HermesChatCrypto.swift
import CryptoKit
import Foundation

public enum HermesChatCryptoError: LocalizedError, Equatable, Sendable {
  case invalidKey
  case invalidEnvelope
  case invalidMessage
  case invalidAttachment
  case decryptionFailed

  public var errorDescription: String? {
    switch self {
    case .invalidKey: "聊天密钥必须是包含 32 字节的 Base64URL。"
    case .invalidEnvelope: "收到不支持的 Hermes 加密格式。"
    case .invalidMessage: "Hermes 消息格式无效。"
    case .invalidAttachment: "Hermes 附件格式无效。"
    case .decryptionFailed: "无法解密 Hermes 消息，请检查此 profile 的聊天密钥。"
    }
  }
}

public struct HermesEncryptedAttachment: Equatable, Sendable {
  public let descriptor: HermesChatAttachmentDescriptor
  public let ciphertext: Data
}

public struct HermesChatCrypto: Sendable {
  public static let maximumAttachmentBytes = 10 * 1024 * 1024
  public static let maximumAgentAttachmentBytes = 512 * 1024 * 1024
  public static let privateAttachmentStorage = "r2-private-v1"

  private let key: SymmetricKey

  public init(encodedKey: String) throws {
    guard let raw = Data(base64URLEncoded: encodedKey), raw.count == 32 else {
      throw HermesChatCryptoError.invalidKey
    }
    key = SymmetricKey(data: raw)
  }

  public static func generateKey() -> String {
    let key = SymmetricKey(size: .bits256)
    return key.withUnsafeBytes { Data($0).base64URLEncodedString() }
  }

  func encryptMessageEnvelope(
    sender: String,
    text: String,
    attachments: [HermesChatAttachmentDescriptor],
    messageID: String,
    sentAt: Int64,
    replaceSequence: Int64 = 0,
    finalUpdate: Bool = false
  ) throws -> HermesChatMessageEnvelope {
    guard sender == "client" || sender == "agent", Self.isCanonicalUUID(messageID) else {
      throw HermesChatCryptoError.invalidMessage
    }
    guard replaceSequence >= 0 else { throw HermesChatCryptoError.invalidMessage }
    for attachment in attachments {
      try Self.validate(descriptor: attachment, sender: sender)
    }
    let payload = HermesChatPlaintextPayload(
      text: text,
      attachments: attachments,
      replaceSeq: replaceSequence > 0 ? replaceSequence : nil,
      final: replaceSequence > 0 ? finalUpdate : nil
    )
    return try encryptPayload(payload, sender: sender, messageID: messageID, sentAt: sentAt)
  }

  func encryptInteractionEnvelope(
    promptID: String,
    optionID: String,
    messageID: String,
    sentAt: Int64
  ) throws -> HermesChatMessageEnvelope {
    guard Self.isCanonicalUUID(promptID), Self.isCanonicalUUID(messageID),
      Self.isInteractionToken(optionID)
    else { throw HermesChatCryptoError.invalidMessage }
    return try encryptPayload(
      HermesChatPlaintextPayload(
        text: "",
        interactionResponse: HermesChatInteractionResponse(
          promptId: promptID,
          optionId: optionID
        )
      ),
      sender: "client",
      messageID: messageID,
      sentAt: sentAt
    )
  }

  func decryptMessageEnvelope(
    _ envelope: HermesChatMessageEnvelope,
    sequence: Int64
  ) throws -> HermesChatMessage {
    guard envelope.v == 1, envelope.type == "message",
      envelope.sender == "client" || envelope.sender == "agent",
      Self.isCanonicalUUID(envelope.id), envelope.sentAt >= 0,
      envelope.encrypted.alg == "A256GCM"
    else { throw HermesChatCryptoError.invalidEnvelope }
    do {
      let plaintext = try open(
        encrypted: envelope.encrypted,
        authenticating: Self.messageAAD(
          id: envelope.id,
          sender: envelope.sender,
          sentAt: envelope.sentAt
        )
      )
      let payload = try JSONDecoder().decode(HermesChatPlaintextPayload.self, from: plaintext)
      let replaceSequence = payload.replaceSeq ?? 0
      let finalUpdate = payload.final ?? false
      if (payload.replaceSeq == nil) != (payload.final == nil) || replaceSequence < 0 {
        throw HermesChatCryptoError.invalidMessage
      }
      if let interaction = payload.interaction {
        try Self.validate(interaction: interaction, sender: envelope.sender)
      }
      guard payload.text.count <= 50_000, payload.attachments.count <= 8 else {
        throw HermesChatCryptoError.invalidMessage
      }
      for attachment in payload.attachments {
        try Self.validate(descriptor: attachment, sender: envelope.sender)
      }
      return HermesChatMessage(
        id: envelope.id,
        sender: envelope.sender,
        sentAt: envelope.sentAt,
        sequence: max(0, sequence),
        text: payload.text,
        attachments: payload.attachments,
        interaction: payload.interaction,
        replaceSequence: replaceSequence,
        finalUpdate: finalUpdate
      )
    } catch let error as HermesChatCryptoError {
      throw error
    } catch {
      throw HermesChatCryptoError.decryptionFailed
    }
  }

  public func encryptAttachment(
    _ plaintext: Data,
    contentType: String,
    filename: String
  ) throws -> HermesEncryptedAttachment {
    guard !plaintext.isEmpty, plaintext.count <= Self.maximumAttachmentBytes else {
      throw HermesChatCryptoError.invalidAttachment
    }
    let id = UUID().uuidString.lowercased()
    let safeType = try Self.safeContentType(contentType)
    let nonce = AES.GCM.Nonce()
    let sealed = try AES.GCM.seal(
      plaintext,
      using: key,
      nonce: nonce,
      authenticating: Self.attachmentAAD(
        id: id,
        contentType: safeType,
        bytes: plaintext.count
      )
    )
    let descriptor = HermesChatAttachmentDescriptor(
      id: id,
      name: Self.safeFilename(filename),
      contentType: safeType,
      plaintextBytes: plaintext.count,
      nonce: Data(nonce).base64URLEncodedString()
    )
    return HermesEncryptedAttachment(
      descriptor: descriptor,
      ciphertext: sealed.ciphertext + sealed.tag
    )
  }

  public func decryptAttachment(
    _ ciphertext: Data,
    descriptor: HermesChatAttachmentDescriptor
  ) throws -> Data {
    try Self.validate(descriptor: descriptor, sender: "client")
    guard let nonce = descriptor.nonce,
      let nonceData = Data(base64URLEncoded: nonce), nonceData.count == 12,
      ciphertext.count >= 16
    else { throw HermesChatCryptoError.invalidAttachment }
    do {
      let sealed = try AES.GCM.SealedBox(
        nonce: AES.GCM.Nonce(data: nonceData),
        ciphertext: ciphertext.dropLast(16),
        tag: ciphertext.suffix(16)
      )
      let plaintext = try AES.GCM.open(
        sealed,
        using: key,
        authenticating: Self.attachmentAAD(
          id: descriptor.id,
          contentType: descriptor.contentType,
          bytes: descriptor.plaintextBytes
        )
      )
      guard plaintext.count == descriptor.plaintextBytes else {
        throw HermesChatCryptoError.invalidAttachment
      }
      return plaintext
    } catch let error as HermesChatCryptoError {
      throw error
    } catch {
      throw HermesChatCryptoError.decryptionFailed
    }
  }

  func encryptLocalSnapshot(_ plaintext: Data, spaceID: String) throws -> HermesChatLocalEnvelope {
    let nonce = AES.GCM.Nonce()
    let sealed = try AES.GCM.seal(
      plaintext,
      using: key,
      nonce: nonce,
      authenticating: try Self.localSnapshotAAD(spaceID: spaceID)
    )
    return HermesChatLocalEnvelope(
      v: 1,
      alg: "A256GCM",
      nonce: Data(nonce).base64URLEncodedString(),
      ciphertext: (sealed.ciphertext + sealed.tag).base64URLEncodedString()
    )
  }

  func decryptLocalSnapshot(_ envelope: HermesChatLocalEnvelope, spaceID: String) throws -> Data {
    guard envelope.v == 1, envelope.alg == "A256GCM",
      let nonce = Data(base64URLEncoded: envelope.nonce), nonce.count == 12,
      let combined = Data(base64URLEncoded: envelope.ciphertext), combined.count >= 16
    else { throw HermesChatCryptoError.invalidEnvelope }
    do {
      let box = try AES.GCM.SealedBox(
        nonce: AES.GCM.Nonce(data: nonce),
        ciphertext: combined.dropLast(16),
        tag: combined.suffix(16)
      )
      return try AES.GCM.open(
        box,
        using: key,
        authenticating: Self.localSnapshotAAD(spaceID: spaceID)
      )
    } catch let error as HermesChatCryptoError {
      throw error
    } catch {
      throw HermesChatCryptoError.decryptionFailed
    }
  }

  private func encryptPayload(
    _ payload: HermesChatPlaintextPayload,
    sender: String,
    messageID: String,
    sentAt: Int64
  ) throws -> HermesChatMessageEnvelope {
    let plaintext = try JSONEncoder().encode(payload)
    let nonce = AES.GCM.Nonce()
    let sealed = try AES.GCM.seal(
      plaintext,
      using: key,
      nonce: nonce,
      authenticating: Self.messageAAD(id: messageID, sender: sender, sentAt: sentAt)
    )
    return HermesChatMessageEnvelope(
      v: 1,
      type: "message",
      id: messageID,
      sender: sender,
      sentAt: sentAt,
      encrypted: HermesChatEncryptedFields(
        alg: "A256GCM",
        nonce: Data(nonce).base64URLEncodedString(),
        ciphertext: (sealed.ciphertext + sealed.tag).base64URLEncodedString()
      )
    )
  }

  private func open(encrypted: HermesChatEncryptedFields, authenticating aad: Data) throws -> Data {
    guard let nonce = Data(base64URLEncoded: encrypted.nonce), nonce.count == 12,
      let combined = Data(base64URLEncoded: encrypted.ciphertext), combined.count >= 16
    else { throw HermesChatCryptoError.invalidEnvelope }
    let box = try AES.GCM.SealedBox(
      nonce: AES.GCM.Nonce(data: nonce),
      ciphertext: combined.dropLast(16),
      tag: combined.suffix(16)
    )
    return try AES.GCM.open(box, using: key, authenticating: aad)
  }

  private static func messageAAD(id: String, sender: String, sentAt: Int64) -> Data {
    Data("hermes-chat-message-v1|\(id)|\(sender)|\(sentAt)".utf8)
  }

  private static func attachmentAAD(id: String, contentType: String, bytes: Int) -> Data {
    Data("hermes-chat-attachment-v1|\(id)|\(contentType)|\(bytes)".utf8)
  }

  private static func localSnapshotAAD(spaceID: String) throws -> Data {
    let normalized = spaceID.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    guard normalized.range(of: "^[a-z0-9][a-z0-9_-]{0,63}$", options: .regularExpression)
      == normalized.startIndex..<normalized.endIndex
    else { throw HermesChatCryptoError.invalidMessage }
    return Data("hermes-chat-local-cache-v1|\(normalized)".utf8)
  }

  private static func validate(
    descriptor: HermesChatAttachmentDescriptor,
    sender: String
  ) throws {
    guard isCanonicalUUID(descriptor.id), descriptor.plaintextBytes > 0,
      !descriptor.name.isEmpty, descriptor.name.count <= 120,
      (try? safeContentType(descriptor.contentType)) == descriptor.contentType
    else { throw HermesChatCryptoError.invalidAttachment }
    if descriptor.isPrivateR2 {
      guard sender == "agent",
        descriptor.plaintextBytes > maximumAttachmentBytes,
        descriptor.plaintextBytes <= maximumAgentAttachmentBytes,
        descriptor.nonce?.isEmpty != false,
        let sha256 = descriptor.sha256,
        sha256.count == 64,
        sha256 == sha256.lowercased(),
        sha256.range(of: "^[0-9a-f]{64}$", options: .regularExpression)
          == sha256.startIndex..<sha256.endIndex
      else { throw HermesChatCryptoError.invalidAttachment }
      return
    }
    guard descriptor.storage == nil || descriptor.storage == "" || descriptor.storage == "encrypted-v1",
      descriptor.sha256 == nil || descriptor.sha256 == "",
      descriptor.plaintextBytes <= maximumAttachmentBytes,
      let nonce = descriptor.nonce,
      Data(base64URLEncoded: nonce)?.count == 12
    else { throw HermesChatCryptoError.invalidAttachment }
  }

  private static func validate(interaction: HermesChatInteraction, sender: String) throws {
    let kinds = ["approval", "slash_confirm", "clarify", "model", "choice"]
    let states = ["pending", "resolved", "expired", "error"]
    guard sender == "agent", isCanonicalUUID(interaction.id), kinds.contains(interaction.kind),
      states.contains(interaction.state), interaction.expiresAt >= 0,
      interaction.options.count <= 24,
      (interaction.state != "pending" || !interaction.options.isEmpty),
      (interaction.stage?.count ?? 0) <= 32,
      (interaction.status?.count ?? 0) <= 500,
      (interaction.pageInfo?.count ?? 0) <= 100
    else { throw HermesChatCryptoError.invalidMessage }
    for option in interaction.options {
      guard isInteractionToken(option.id), !option.label.isEmpty, option.label.count <= 120,
        ["default", "primary", "danger", "warning"].contains(option.style)
      else { throw HermesChatCryptoError.invalidMessage }
    }
  }

  private static func isCanonicalUUID(_ value: String) -> Bool {
    guard let parsed = UUID(uuidString: value) else { return false }
    return parsed.uuidString.lowercased() == value.lowercased() && value == value.lowercased()
  }

  private static func isInteractionToken(_ value: String) -> Bool {
    guard (1...64).contains(value.count) else { return false }
    return value.range(of: "^[A-Za-z0-9_-]+$", options: .regularExpression)
      == value.startIndex..<value.endIndex
  }

  private static func safeContentType(_ value: String) throws -> String {
    let normalized = value.split(separator: ";", maxSplits: 1).first.map(String.init)?
      .trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
      ?? "application/octet-stream"
    guard !normalized.isEmpty, normalized.count <= 100,
      normalized.range(of: "^[a-z0-9!#$&^_.+\\-/]+$", options: .regularExpression)
        == normalized.startIndex..<normalized.endIndex
    else { throw HermesChatCryptoError.invalidAttachment }
    return normalized
  }

  private static func safeFilename(_ value: String) -> String {
    let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "._ -"))
    let output = value.unicodeScalars.map { allowed.contains($0) ? Character(String($0)) : "_" }
    var filename = String(output).trimmingCharacters(in: CharacterSet(charactersIn: " ."))
    if filename.isEmpty { filename = "attachment" }
    return String(filename.prefix(120))
  }
}

private extension Data {
  init?(base64URLEncoded value: String) {
    var normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
      .replacingOccurrences(of: "-", with: "+")
      .replacingOccurrences(of: "_", with: "/")
    let remainder = normalized.count % 4
    if remainder != 0 { normalized.append(String(repeating: "=", count: 4 - remainder)) }
    self.init(base64Encoded: normalized)
  }

  func base64URLEncodedString() -> String {
    base64EncodedString()
      .replacingOccurrences(of: "+", with: "-")
      .replacingOccurrences(of: "/", with: "_")
      .replacingOccurrences(of: "=", with: "")
  }
}
